controller/volume_controller.go

```diff
@@ -3005,6 +3006,138 @@ func (vc *VolumeController) ResolveRefAndEnqueue(namespace string, ref *metav1.O
 	vc.enqueueVolume(volume)
 }
 
+// createNewRecurringJobNameUsed use a new name to create a new recurring job that original name exists but configuration is different
+func createNewRecurringJobNameUsed(log *logrus.Entry, ds *datastore.DataStore, job *longhorn.RecordVolumeRecurringJob) (string, error) {
+	// reuse restored recurring job that RecurringJobSpec is the same.
+	allRecurringJobs, err := ds.ListRecurringJobsRO()
+	if err != nil {
+		log.WithError(err).Warn("Failed to list all recurring jobs")
+	}
+	for existJobName, existJob := range allRecurringJobs {
+		if !strings.HasPrefix(existJobName, types.VolumeRecurringJobRestorePrefix) {
+			continue
+		}
+		job.JobSpec.Name = existJob.Spec.Name
+		if reflect.DeepEqual(existJob.Spec, job.JobSpec) {
+			return existJobName, nil
+		}
+	}
+
+	newJobName := types.VolumeRecurringJobRestorePrefix + util.RandomID() + util.RandomID()
+	oldJobName := job.JobSpec.Name
+	job.JobSpec.Name = newJobName
+
+	log.Debugf("Create a new recurring job %v instead of a existing job name %v when restoring recurring jobs", newJobName, oldJobName)
+	if _, err := ds.CreateRecurringJob(&longhorn.RecurringJob{ObjectMeta: metav1.ObjectMeta{Name: newJobName}, Spec: job.JobSpec}); err != nil {
+		log.WithError(err).Warnf("Failed to create a new recurring job %v for restoring the job %v", newJobName, oldJobName)
+		return "", err
+	}
+
+	return newJobName, nil
+}
+
+// createRecurringJobNotExistAndLabel add recurring jobs and group back to a restoring volume and create recurring jobs if not exist
+func createRecurringJobNotExistAndLabel(log *logrus.Entry, ds *datastore.DataStore, v *longhorn.Volume, jobName string, job *longhorn.RecordVolumeRecurringJob) error {
+	if existJob, err := ds.GetRecurringJob(jobName); err != nil {
+		if !apierrors.IsNotFound(err) {
+			log.WithError(err).Warnf("Failed to get the recurring job %v", jobName)
+			return err
+		}
+		// create the recurring job if not exist
+		log.Debugf("Create the recurring job %v when restoring recurring jobs", jobName)
+		if _, err := ds.CreateRecurringJob(&longhorn.RecurringJob{ObjectMeta: metav1.ObjectMeta{Name: jobName}, Spec: job.JobSpec}); err != nil {
+			log.WithError(err).Warnf("Failed to create a recurring job %v", jobName)
+			return err
+		}
+	} else if !reflect.DeepEqual(existJob.Spec, job.JobSpec) {
+		// create a new recurring job if job name is used and configuration is different.
+		newJobName, err := createNewRecurringJobNameUsed(log, ds, job)
```

Comment:
Since you already have a `ds.CreateRecurringJob` in this function, suggest `createNewRecurringJobNameUsed` can return the RecurringJob obj and have it create here in the same function.

Reply:
Modified. Thanks

Reply:
I mean you can `ds.CreateRecurringJob` in the caller(`createRecurringJobNotExist`).

Reply:
OK, `createNewRecurringJobNameUsed` is only used to create a new name or find a restored reuccuring jobs that its spec is the same to the job user want to restore.
And create recurring job only in `createRecurringJobNotExist` if needed.